File: python/Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        if not head or not head.next:
            return head

        answer = head.next
        cursor = head
        prev = None
        
        while cursor and cursor.next:
            tmp = cursor.next
            cursor.next = cursor.next.next
            tmp.next = cursor
            if prev: prev.next = tmp
            prev = cursor
            cursor = cursor.next
        
        return answer

File: cpp/Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        ListNode *cur = head, *prev = NULL;
        ListNode *answer = head->next;
        while (cur && cur->next) {
            ListNode *curNext = cur->next;
            cur->next = cur->next->next;
            curNext->next = cur;
            if (prev) {
                prev->next = curNext;    
            }
            prev = cur;
            cur = cur->next;
        }
        return answer;
    }
};
